[BE] Password update

nspc97 opened the pull request:
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

Changed register and user_update logic.

Commit:
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

Changed register and user_update logic.

Commit:
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -89,16 +90,16 @@ public function editUser($id, SerializerInterface $serializer, Request $request,
         $data = $request->getContent();
```
We forgot to add a check that you can edit only your profile. Please compare the logged in user with the user which is being edited.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -145,4 +147,59 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
+        if (!$user) {
+            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+        }
+        $data = $request->getContent();
```
Add a check that you can change only your password. Please compare the logged in user with the user which is being edited.

SpartakusMd commented on src/DTO/UserDTO.php:
```diff
@@ -16,86 +20,150 @@ class UserDTO
+     * @Serializer\Expose()
+     * @Groups({"PasswordEdit"})
+     */
+    public $confirmNewPassword;
```
You can rename the fields so you don't have 2 confirm fields. On register you would have password, confirmPassword and on password change you would have oldPassword, password, confirmPassword. The validation rules will also be simpler in this case.

SpartakusMd commented on src/Exceptions/NotValidOldPassword.php:
```diff
@@ -0,0 +1,11 @@
+    {
+        parent::__construct($message);
+    }
+}
```
Add new lines at the end of files.

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -90,4 +104,33 @@ public function transform(
+        $entity->setPassword($this->encoder->encodePassword($entity, $dto->newPassword));
+        return $entity;
+    }
+}
```
New line.

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -90,4 +104,33 @@ public function transform(
+    public function changePasswordTransform(
+        UserDTO $dto
+    ): User {
+        $entity = $this->userRepo->find($dto->id);
```
On change password you should have only the passwords in the requests. The user id is sent in URL. Pass the $user object when calling `changePasswordTransform()`

MihaiDruta commented on src/Controller/RegisterController.php:
```diff
@@ -57,7 +57,7 @@ public function createUser(SerializerInterface $serializer, Request $request, Us
-        $context = DeserializationContext::create();
+        $context = DeserializationContext::create()->setGroups(array('UserCreate'));
 
         $userDTO = $serializer->deserialize(
```
Inject the serializer in `__construct()` method like `validator` and `transformer`

Commit:
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

Added check constraints for edit user, change password, for only your account.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
      * @throws OptimisticLockException
      */
-    public function editUser($id, Request $request, UserRepository $userRepository)
+    public function editUser($id, Request $request, UserRepository $userRepository, UserInterface $user)
```
I would use `$currentUser` or `$loggedUser` or `$authenticatedUser` as it better reflects what kind of user is this.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
-        if (!$user) {
-            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+        $userUsername = $user->getUsername();
+        $userPage = $userRepository->find($id);
```
This is a user, not a page. Use `$user` or `$affectedUser`.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
 
-        $data = $request->getContent();
+        $userPage = $userRepository->find($id)->getUsername();
+        if ($userUsername === $userPage) {
```
Compare user ids, not usernames ..

SpartakusMd replied:
It should be the reverse. If the ids (or usernames in your case) don't match, return the error JSON, as so, you won't have too much identation and you should alway fail / return fast from the functions.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -151,55 +157,66 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
+        $id,
+        UserRepository $userRepository,
+        Request $request,
+        UserInterface $user
```
Same here, rename `$user`

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -40,39 +40,54 @@ public function __construct(
+     */
+    public function editTransform(
+        UserDTO $dto,
+        $id
```
Send the User object, not the id. You already searched the user in the database, there's no reason to do it twice.

SpartakusMd commented on src/Service/UserTransformer.php:
```diff
@@ -90,4 +105,34 @@ public function transform(
+     */
+    public function changePasswordTransform(
+        UserDTO $dto,
+        $id
```
Send the User object, not the id. You already searched the user in the database, there's no reason to do it twice.

Commit:
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

Added check constraints for edit user, change password, for only your account.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -70,45 +71,43 @@ public function getUserDetails(User $user): Response
+        $authenticatedUser = $this->getUser();
+
+        if ($authenticatedUser->getId() !== $user->getId()) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_BAD_REQUEST);
```
Actualy it should be `Response::HTTP_FORBIDDEN` as the current user cannot edit this user. You can fix it after the merge.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -145,4 +144,53 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
+        $authenticatedUser = $this->getUser();
+
+        if ($authenticatedUser->getId() !== $user->getId()) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_BAD_REQUEST);
```
Actualy it should be Response::HTTP_FORBIDDEN as the current user cannot edit this user. You can fix it after the merge.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -145,4 +144,53 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
+        if ($authenticatedUser->getId() !== $user->getId()) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_BAD_REQUEST);
+        }
+            $data = $request->getContent();
```
Indentation is wrong in this method :)

SpartakusMd commented on src/Exceptions/NotValidOldPassword.php:
```diff
@@ -0,0 +1,11 @@
+
+class NotValidOldPassword extends \Exception
+{
+    public function __construct(string $message = '')
```
Remove the construct if it doesn't do anything.
